graphs/possible_bipartition.py

```diff
@@ -1,12 +1,46 @@
-# Can be used for BFS
-from collections import deque 
+from collections import defaultdict  
 
 def possible_bipartition(dislikes):
     """ Will return True or False if the given graph
         can be bipartitioned without neighboring nodes put
         into the same partition.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(V+E)
+        Space Complexity: O(V+E)
     """
-    pass
+    # create an adjacency list to represent the graph
+    graph = defaultdict(list)
+    for dislike in dislikes:
+        # if there are no enemies, no need to add to graph
+        if len(dislike) > 1:
+            dog, enemy = dislike[0], dislike[1]
+            graph[dog].append(enemy)
+        # if there are 3 or more dogs who fight each other, 
+        # you can't divide the non-fighters into just 2 groups,
+        # so return False immediately 
+        if len(dislike) > 2:
+            return False 
+    
+    # create a hash where dogs will be assigned to group 1 or group -1
+    grouped = {}
+    
+    for dog in range(len(graph)):
+        if dog not in grouped:
+            grouped[dog] = 1
+        # do BFS 
+        queue = [dog]
+        while queue:
+            curr_dog = queue.pop(0)
+            for enemy in graph[curr_dog]:
+                if enemy not in grouped:
```

Comment:
🌸 Very neat to use `grouped` to track visited dogs/nodes as well as their group!